dGame/dCinema/Recorder.cpp

```diff
@@ -0,0 +1,1195 @@
+#include "Recorder.h"
+
+#include "ControllablePhysicsComponent.h"
+#include "GameMessages.h"
+#include "InventoryComponent.h"
+#include "../dWorldServer/ObjectIDManager.h"
+#include "ChatPackets.h"
+#include "EntityManager.h"
+#include "EntityInfo.h"
+#include "ServerPreconditions.hpp"
+#include "MovementAIComponent.h"
+#include "BaseCombatAIComponent.h"
+
+using namespace Cinema::Recording;
+
+std::unordered_map<LWOOBJID, Recorder*> m_Recorders = {};
+
+std::unordered_map<int32_t, std::string> m_EffectAnimations = {};
+
+Recorder::Recorder() {
+	this->m_StartTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
+	this->m_IsRecording = false;
+	this->m_LastRecordTime = this->m_StartTime;
+}
+
+Recorder::~Recorder() {
+}
+
+void Recorder::AddRecord(Record* record)
+{
+	if (!this->m_IsRecording) {
+		return;
+	}
+
+	LOG("Adding record");
+
+	const auto currentTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
+
+	// Time since the last record time in seconds
+	record->m_Delay = (currentTime - this->m_LastRecordTime).count() / 1000.0f;
+
+	m_LastRecordTime = currentTime;
+
+	this->m_Records.push_back(record);
+}
+
+void Recorder::Act(Entity* actor, Play* variables) {
+	LOG("Acting %d steps", m_Records.size());
+
+	// Loop through all records
+	ActingDispatch(actor, m_Records, 0, variables);
+}
+
+void Recorder::ActingDispatch(Entity* actor, const std::vector<Record*>& records, size_t index, Play* variables) {
+	if (index >= records.size()) {
+		return;
+	}
+
+	auto* record = records[index];
+
+	// Check if the record is a fork
+	auto* forkRecord = dynamic_cast<ForkRecord*>(record);
```

Comment:
I think I see what is going on here; I would recommend moving these into handler functions called Dispatch for each of the record types and have them overload what needs to be done based on that instead of a bunch of dynamic casts.  Would make a much smaller function and split functionality up for better maintainability